src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1926,4 +1942,27 @@ public static String getVersionName() {
         // value
         return "1.1.12";
     }
+
+    /**
+     * A {@link Future}  whose result can be set by a {@link #set(Object)} or {@link #setException(Throwable)}
+     */
+    final static class SettableFuture<V> extends FutureTask<V> {
```

Comment:
I have to warn about the problem that this solution can cause.
The main concern is method   
public Future<AuthenticationResult> acquireTokenSilent(String resource, String clientId,
            String userId, AuthenticationCallback<AuthenticationResult> callback).
First I don't really like it's public, why caller would ever need Future and deal with it.... I think it meant to be private. But anyway it's already public and as I understand we want to support it.

So the problem if we throw within afterExecutre AuthenticationException,
even if user calls Future.get() and than catches AuthenticationException we still crash...
I guess it might be ok, since if we throw AuthenticationException instead of returning them in callback we treat them as IlligalStateException....But I want you to know about this consequence....